src/main/java/org/opensearch/knn/training/TrainingJobRunner.java

```diff
@@ -176,4 +193,106 @@ private void serializeModel(TrainingJob trainingJob, ActionListener<IndexRespons
     public int getJobCount() {
         return jobCount.get();
     }
+
+    /**
+     * This method is called whenever the cluster state changes. It is used to update models that are still training when a node leaves or the cluster crashes.
+     * @param event the event that changed the cluster change
+     */
+    @Override
```

Comment:
I think we probably will want a separate class for this logic. The TrainingJobRunner is mostly meant to run the job itself - so it shouldnt ever be running on the cluster manager node. What do you think?

Reply:
I think we should keep it in this class, since it originally handled all of the error/exception logic for when training failed.  In my opinion, this is just an extension of the job failing, so the class running the job should handle that failure.

Reply:
The issue is that this class is responsible for executing the job on the node. It is not responsible for maintaining the cluster state. So, the cluster manager node will need to run this operation, but it will never use any functionality outside of this.

This then couples the logic used to maintain cluster state with the logic to submit the actual training job. The problem with this is that it makes it more difficult to test because it adds dependencies to the class. Instead, we could just have a separate `ClusterStateListener` with the sole responsibility of marking models as failed when the node drops. Something like `TrainingJobClusterStateListener` that manages the cluster state for training jobs.

Reply:
Ok I'll try to refactor into a new class